app/models/user.rb

```diff
@@ -4,10 +4,40 @@ class User < ApplicationRecord
   devise :database_authenticatable, :registerable,
          :recoverable, :rememberable, :trackable, :validatable
 
+  # Virtual attribute for authenticating by either username or email
+  # This is in addition to a real persisted field like 'username'
+  attr_accessor :login
+
+  validates_format_of :username, with: /^[a-zA-Z0-9_\.]*$/, multiline: true
+  validates :username,
+  presence: true,
+  uniqueness: {
+    case_sensitive: false
+  }
+
   has_one :homepage, dependent: :destroy
   after_create -> { self.create_homepage }
 
   has_many :lists
 
   acts_as_voter
+
+  def self.find_for_database_authentication(warden_conditions)
+    conditions = warden_conditions.dup
+    if login = conditions.delete(:login)
+      where(conditions.to_h).where(["lower(username) = :value OR lower(email) = :value", { :value => login.downcase }]).first
+    elsif conditions.has_key?(:username) || conditions.has_key?(:email)
+      conditions[:email].downcase! if conditions[:email]
+      where(conditions.to_h).first
+    end
```

Comment:
Oh, this is at a model level, not for the forms. So `login` remains the only parameter you can authenticate by.

But it does look like it downcases the email and not the username, if one is provided. I would want to downcase the username as well, but I'm not sure when and where `find_for_database_authentication` is used.